### PR TITLE
feat: faster blt indexing via known rectangularity

### PR DESCRIPTION
**Pull request by steven-murray**
## Description

This improves performance for `antpair2ind()` and `_key2inds` in a couple of ways:

1. It caches the results, with some extra logic in there for better cache-invalidation
2. It uses the `blt_ordering` and whether the blts are "rectangular" to enable faster indexing if relevant. These functions now return slices wherever possible, instead of int arrays, which cuts down on memory usage (and is much faster to index by). This has necessitated some fixing up of tests that used the outputs of these assuming they are numpy arrays.

Other than these results being slightly different, there should be no other breaking changes here,  just extra speed if you want it.

## Motivation and Context

Indexing the data can be slow, this should speed it up. I'll report back with performance metrics soon.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation change (documentation changes only)
- [ ] Version change
- [ ] Build or continuous integration change

Performance enhancement?

## Checklist:

- [x] I have read the [contribution guide](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/main/.github/CONTRIBUTING.md).
- [x] My code follows the code style of this project.

New feature checklist:
- [x] I have added or updated the docstrings associated with my feature using the [numpy docstring format](https://numpydoc.readthedocs.io/en/latest/format.html).
- [ ] I have updated the tutorial to highlight my new feature (if appropriate).
- [x] I have added tests to cover my new feature.
- [x] All new and existing tests pass.
- [x] I have updated the [CHANGELOG](https://github.com/RadioAstronomySoftwareGroup/pyuvdata/blob/main/CHANGELOG.md).

### CODE REVIEW

File: pyuvdata/uvdata/uvdata.py [hunk 1]
```diff
@@ -34,6 +34,7 @@
 logger = logging.getLogger(__name__)
 
 allowed_cat_types = ["sidereal", "ephem", "unprojected", "driftscan"]
+EMPTY_INDX = np.array([], dtype=np.int64)
```
Comment (bhazelton) [inline pyuvdata/uvdata/uvdata.py:37]: Doesn't look like this is used anywhere.

Comment (steven-murray) [reply pyuvdata/uvdata/uvdata.py]: Removed.

File: pyuvdata/uvdata/uvdata.py [hunk 2]
```diff
@@ -3715,9 +3777,9 @@ def _key2inds(self, key):
 
         Returns
         -------
-        blt_ind1 : ndarray of int
-            blt indices for antenna pair.
-        blt_ind2 : ndarray of int
+        blt_ind1 : ndarray of int or slice or None
+            blt indices for antenna pair. None if antpair is not in data.
```
Comment (bhazelton) [inline pyuvdata/uvdata/uvdata.py:3781]: It looks like it raises an error if the antpair isn't in the data? Can you add a `Raises` section to this docstring?

File: CHANGELOG.md [hunk 3]
```diff
@@ -16,6 +16,8 @@ positions are near surface of whatever celestial body their positions are refere
 (either the Earth or Moon, currently).
 
 ### Changed
+- Improved performance for `antpair2ind` and `_key2inds` by using the
+`blts_are_rectangular` parameter, and also by caching the results. To improve performance of the cache, the resulting indices are returned as slices whenever possible.
```
Comment (bhazelton) [inline CHANGELOG.md:20]: Minor quibble, but can you wrap this line to make it similar to the rest of the file?

File: pyuvdata/uvdata/uvdata.py [hunk 4]
```diff
@@ -4793,6 +4754,9 @@ def conjugate_bls(self, convention="ant1<ant2", *, use_enu=True, uvw_tol=0.0):
             )
             self.Nbls = np.unique(self.baseline_array).size
 
+            self.__antpair2ind_cache = {}
+            self._key2ind_cache = {}
```
Comment (bhazelton) [inline pyuvdata/uvdata/uvdata.py:4758]: Why not just call the `_clear_antpair2ind_cache` method here?

File: pyuvdata/uvdata/tests/test_uvdata.py [hunk 5]
```diff
@@ -4964,374 +4975,189 @@ def test_smart_slicing_err(casa_uvfits):
...
+REG_POL = slice(0, 2)
+IRREG_IND = [0, 4, 5]
+IRREG_POL = [0, 1, 3]
+
+
 @pytest.mark.filterwarnings("ignore:Telescope EVLA is not")
 @pytest.mark.filterwarnings("ignore:The uvw_array does not match the expected values")
-@pytest.mark.parametrize("future_shapes", [True, False])
-def test_smart_slicing(casa_uvfits, future_shapes):
+@pytest.mark.parametrize(
```
Comment (bhazelton) [inline pyuvdata/uvdata/tests/test_uvdata.py:4995]: Thank you for parametrizing this test!!

File: pyuvdata/uvdata/tests/test_uvdata.py [hunk 6]
```diff
@@ -4964,374 +4975,189 @@ def test_smart_slicing_err(casa_uvfits):
...
+
+        uv.data_array[bltinds[1], 0, polidx[0]] = 5.43
+        if copy_made:
+            assert d[1, 0, 0] != uv.data_array[bltinds[1], 0, polidx[0]]
+        else:
+            assert d[1, 0, 0] == uv.data_array[bltinds[1], 0, polidx[0]]
 
 
-def test_get_data(casa_uvfits):
+@pytest.mark.parametrize("kind", ["data", "flags", "nsamples", "times", "lsts"])
```
Comment (bhazelton) [inline pyuvdata/uvdata/tests/test_uvdata.py:5077]: And this one!

File: CHANGELOG.md (hunk 3 above)
Comment (steven-murray) [reply CHANGELOG.md:20]: Ah, my editor wrapped it for me so I didn't notice. Thanks!

File: pyuvdata/utils.py [hunk 7]
```diff
@@ -6285,3 +6285,16 @@ def determine_rectangularity(
         if (np.diff(baseline_array, axis=0) != 0).any():
             return False, False
         return True, False
+
+
+def slicify(ind):
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply pyuvdata/utils.py:6290]: Oh yeah. Done.

File: pyuvdata/uvdata/uvdata.py (hunk 4 above)
Comment (steven-murray) [reply pyuvdata/uvdata/uvdata.py:4758]: good idea...

File: pyuvdata/uvdata/uvdata.py (hunk 2 above)
Comment (steven-murray) [reply pyuvdata/uvdata/uvdata.py:3781]: Ah yeah this was a bit confusing. Behaviour is that if the baseline in a particular direction (e.g. (1,0,'xx')) does not exist, it returns None for that direction, but if they BOTH don't exist (so also (0, 1, 'xx') doesn't exist) then an error is raised. I added a Raises section as you suggest.

File: pyuvdata/uvdata/uvdata.py [hunk 8]
```diff
@@ -3640,7 +3657,7 @@ def antnums_to_baseline(self, ant1, ant2, *, attempt256=False):
             ant1, ant2, Nants_telescope=self.Nants_telescope, attempt256=attempt256
         )
 
-    def antpair2ind(self, ant1, ant2=None, *, ordered=True):
+    def antpair2ind(self, ant1, ant2=None, *, ordered=True) -> np.ndarray | slice:
```
[the remark this answers is not in the document]
Comment (steven-murray) [reply pyuvdata/uvdata/uvdata.py:3660]: Done.

**Review by bhazelton — APPROVED** [review]
Looks good. Let's make issues on `hera_qm` and `hera_cal` to track the need for changes there.

**Pull request merged**